src/sqlfluff/core/parser/segments/base.py

```diff
@@ -391,6 +395,35 @@ def raw(self) -> str:
         """Make a string from the segments of this segment."""
         return "".join(seg.raw for seg in self.segments)
 
+    @cached_property
```

Comment:
See my comment on the `_class_types()` function. It seems like it may be possible to fully precompute this at class definition time.

Reply:
We can precompute it for most, but for segments like `KeywordSegment` we need to be able to add a shim for some instance-specific logic. See my comment in reply to your suggestion of metaclasses.